### PR TITLE
Improve Graph Editor UX

### PR DESCRIPTION
**Pull request by lgeertsen**
# Redesigned Nodes
Inputs and outputs are now ordered vertically. And the inputs and outputs are now separated left and right which makes it easier to visualize the graph

image

When nodes are computing or submitted, all the nodes in that branch are **locked**.
Added a lock icon and on overlay over the attributes.
While locked none of the attributes can be edited and no new inputs can be attached nor deleted.
While locked outputs can still be connected.

image

If a nodes in a branch have the same values as a node in another branch than an icon will be displayed to indicate that this node is a duplicate.
When you hover over the icon a tooltip will be displayed with the name of the original node.

Screenshot_2019-08-28_10-40-54

Nodes that are being computed on a renderfarm will have a cloud icon.

image

# TaskManager
Nodes submitted to a render farm are now also managed by the task manager

+ Added ability to select a node in the task manager 

taskManager

### CODE REVIEW

**Comment by natowi** [comment]
@lgeertsen Good work! How about adding the option to hide (collapse) the inputs/outputs to simplify the node graph and expand them if needed?
collapse

**Comment by natowi** [comment]
Generic node "Input" and "output" should be renamed to their actual input/output type (input/output could be used when a node is collapsed ^).
Also reorder items to reduce crossing connectors.
node3

**Comment by yann-lty** [comment]
Node design has been updated to keep a minimal amount of colors. Attributes high-level labels are used instead of technical names which make it more simple to identify which connections make sense (next step would be strongly typed attribute though).
Edges and connections are also simpler and more explicit. Attributes with connected outputs are now identified.
image

**Comment by natowi** [comment]
btw Is there any meaning to the round and square node connectors?

**Comment by yann-lty** [comment]
Round is for single element, Square is for List. 
You can connect N single elements to a List, one List to a List.
ATM, no visual distinction is made between those two types of connections. This is still WIP, any idea to improve this is welcome :).

**Comment by natowi** [comment]
Here are some improvement ideas: https://github.com/alicevision/meshroom/issues/269

**Comment by natowi** [comment]
@yann-lty 

> Attributes' high-level labels are used instead of technical names, which makes it simpler to identify relevant connections

The new nodes (Panorama*,...) still have the generic "input/output" fields

**Comment by yann-lty** [comment]
Indeed, I have not addressed new nodes coming from the recent merge of develop into this branch. 
A pass of naming uniformization regarding attributes labels should follow.

**Comment by natowi** [comment]
Can we get a little bit more contrast for nodes? Similar to Blender maybe?
blend3

**Comment by julianrendell** [comment]
@yann-lty these changes look really amazing, especially for people new to 'grammatry (like me.)

If you have time, it'd be great if you could add an update as to what's left for this to be completed. If it's really simple, I'd be happy to try and help finish this and see it in the main branch.

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 1]
```diff
@@ -259,9 +251,25 @@ Item {
...
                     point2y: dst.nodeItem.y + dstAnchor.y
                     onPressed: {
-                        if(event.button == Qt.RightButton)
+                        var canEdit = true
+                        if(_reconstruction.computing) {
+                            if(uigraph.taskManager.nodes.contains(edge.src.node)) {
+                                    canEdit = false;
+                            } else {
+                                if(object.globalStatus == "SUCCESS") {
+                                    var nodes = uigraph.graph.onlyNodesFromNode(edge.src.node);
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/GraphEditor.qml:260]: Need to maintain a property on each node for that.

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 2]
```diff
@@ -313,7 +325,27 @@ Item {
...
+                    enabled: {
+                        if(! _reconstruction.computing) {
+                            return true;
+                        }
+
+                        if(uigraph.taskManager.nodes.contains(uigraph.selectedNode)) {
+                                return false;
+                        } else {
+                            if(uigraph.selectedNode.globalStatus == "SUCCESS") {
+                                var nodes = uigraph.graph.onlyNodesFromNode(uigraph.selectedNode);
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/GraphEditor.qml:337]: Need to maintain a property on each node for that.

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 3]
```diff
@@ -329,7 +361,26 @@ Item {
...
+                    enabled: {
+                        if(! _reconstruction.computing) {
+                            return true;
+                        }
+
+                        if(uigraph.taskManager.nodes.contains(uigraph.selectedNode) || ["SUBMITTED", "RUNNING"].includes(_reconstruction.selectedNode.globalStatus)) {
+                                return false;
+                        } else {
+                            if(uigraph.selectedNode.globalStatus == "SUCCESS") {
+                                var nodes = uigraph.graph.onlyNodesFromNode(uigraph.selectedNode);
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/GraphEditor.qml:373]: Need to maintain a property on each node for that.

File: meshroom/ui/qml/GraphEditor/GraphEditor.qml [hunk 4]
```diff
@@ -392,7 +443,26 @@ Item {
...
                     node: object
                     width: uigraph.layout.nodeWidth
-                    readOnly: root.readOnly
+                    readOnly: {
+                        if(! uigraph.computing) {
+                            return false;
+                        }
+
+                        if(object.globalStatus == "SUCCESS") {
+                            var nodes = uigraph.graph.onlyNodesFromNode(object);
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/GraphEditor.qml:452]: Same

File: meshroom/ui/qml/GraphEditor/Node.qml [hunk 5]
```diff
@@ -74,99 +98,183 @@ Item {
...
+                            text: MaterialIcons.layers
+                            font.pointSize: 7
+                            padding: 2
+                            palette.text: Colors.sysPalette.text
+                            ToolTip.text: visible ? "Data has been computed by <b>" + node.nameToLabel(node.chunks.at(0).statusNodeName) + "</b>" : ""
+                        }
+
+                        // Submitted externally indicator
+                        MaterialLabel {
+                            visible: ["SUBMITTED", "RUNNING"].includes(node.globalStatus) && node.chunks.count > 0 && node.chunks.at(0).execModeName === "EXTERN"
```
Comment (fabiencastan) [inline meshroom/ui/qml/GraphEditor/Node.qml:173]: globalExecMode

File: meshroom/ui/qml/main.qml [hunk 6]
```diff
@@ -725,16 +709,53 @@ ApplicationWindow {
...
+                function updateNodeStatus() {
+                    if(! _reconstruction.computing) {
+                        readOnly = false;
+                        return;
+                    }
+
+                    if(node.globalStatus === "SUCCESS") {
+                        var nodes = _reconstruction.graph.onlyNodesFromNode(node);
+                        for(var i = 0; i < nodes.length; i++) {
+                            if(["SUBMITTED", "RUNNING"].includes(nodes[i].globalStatus) && nodes[i].chunks.at(0).statusNodeName == nodes[i].name) {
```
Comment (fabiencastan) [inline meshroom/ui/qml/main.qml:745]: globalStatusNodeName

File: meshroom/ui/qml/main.qml [hunk 7]
```diff
@@ -725,16 +709,53 @@ ApplicationWindow {
...
+                onComputingChanged: { updateNodeStatus() }
+
+                function updateNodeStatus() {
+                    if(! _reconstruction.computing) {
+                        readOnly = false;
+                        return;
+                    }
+
+                    if(node.globalStatus === "SUCCESS") {
+                        var nodes = _reconstruction.graph.onlyNodesFromNode(node);
```
Comment (fabiencastan) [inline meshroom/ui/qml/main.qml:743]: Maintain a property on each node.

**Comment by Julien-Haudegond** [comment]
Work In Progress: avoid long and repetitive process on the QML side to update locked nodes.
- [X] Add a *locked* property to nodes (in core).
- [ ] Make the submission of a node lock every node it depends on.
- [ ] Deal with update when a node is computed (split branches, etc).

File: meshroom/core/attribute.py [hunk 8]
```diff
@@ -172,6 +173,8 @@ def isLink(self):
         # note: directly use self.node.graph._edges to avoid using the property that may become invalid at some point
         return self.node.graph and self.isInput and self in self.node.graph._edges.keys()
 
+        return self in self.node.graph.edges.keys()
```
Comment (fabiencastan) [inline meshroom/core/attribute.py:176]: Duplicate line??

Comment (fabiencastan) [reply meshroom/core/attribute.py:176]: Looks like a merge error:
https://github.com/alicevision/meshroom/commit/fcb2b23c1a6b187d7b184daf237cfa072000af43#diff-22303adc24f17b3d167141163446bbc2R156

**Review by fabiencastan — APPROVED** [review]

**Pull request merged**